src/app/components/top-nav.tsx

```diff
@@ -0,0 +1,72 @@
+"use client";
+import { Button } from "@/components/ui/button";
+import { Separator } from "@/components/ui/separator";
+import {
+  Sheet,
+  SheetContent,
+  SheetHeader,
+  SheetTrigger,
+} from "@/components/ui/sheet";
+import { cn } from "@/lib/utils";
+import { Bars3Icon, HomeIcon } from "@heroicons/react/20/solid";
+import Link, { LinkProps } from "next/link";
+import { usePathname } from "next/navigation";
+import { useState } from "react";
+
+type MobileLinkProps = LinkProps & {
+  children: React.ReactNode;
+  className?: string;
+  onOpenChange: (open: boolean) => void;
+};
+function MobileLink({ onOpenChange, className, ...props }: MobileLinkProps) {
+  const pathname = usePathname();
+  const isActive = pathname === props.href;
+  return (
+    <Link
+      aria-current={isActive ? "page" : undefined}
+      className={cn(
+        "flex flex-row items-center w-full rounded px-4 py-2 aria-[current='page']:bg-gray-100",
+        className
+      )}
+      onClick={() => onOpenChange(false)}
+      {...props}
+    />
+  );
+}
+
+function MobileNav() {
+  const [open, setOpen] = useState(false);
+  return (
+    <Sheet open={open} onOpenChange={setOpen}>
```

Comment:
You don't _have_ to manage the `open` state manually (i.e this can be `uncontrolled`) however because of the way these layouts are nested the TopNav component, and therefore the MobileNav component are always mounted. 

This means that clicking a link doesn't close the Sheet it makes the page navigate under it and then you have to manually close it. 

This is kinda weird experience so here I've chosen to make this state `controlled` and have passed around the `setOpen` function to the link, then in the links `onClick` we call `setOpen(false)` to close the navigation so that the menu closes when you click a link.

Reply:
That all makes sense! For my own understanding I'm trying to trace how open in MobileNav is ever set to true. I see a lot of options to set it to false (and thus close the sidebar). Is it to do with the below? 
`<Sheet open={open} onOpenChange={setOpen}>`
but then onOpenChange returns void?

Reply:
This is where it's a bit complex/hidden. It's managed internally by the `SheetTrigger` component. 

Whenever it is clicked (or the child button is clicked) it sets the sheets internal `open` value to true directly. Then when our links are clicked we set our state to `false` which is then passed into the Sheet, which in turn sets _its_ state to `false`. 

Perhaps for clarity we can manually set this whenever the `SheetTrigger` is clicked 🤔

Reply:
We sure could! But it does work at the moment but in my web search journey to understand what was going on I did see some people who were manually setting open to true in similar contexts and they all seemed to be having issues, so I think let's leave it for now.

Appreciate the explainer, that's good to know :)